### PR TITLE
Download file

### LINKED ISSUE #909: pc.download_files()
We have pc.upload_files() for uploading files. I think we should also have pc.download_files() for downloading files. I don't know of a clean way to create a download link in pynecone. We should have one. #795

### PR DESCRIPTION
**Pull request by TaiJuWu**
### All Submissions:

- [X] Have you followed the guidelines stated in [CONTRIBUTING.md](https://github.com/pynecone-io/pynecone/blob/main/CONTRIBUTING.md) file?
- [X] Have you checked to ensure there aren't any other open [Pull Requests](https://github.com/pynecone-io/pynecone/pulls ) for the desired changed?

### Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

### New Feature Submission:

- [X] Does your submission pass the tests? 
- [X] Have you linted your code locally prior to submission?

### Changes To Core Features:

- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [X] Have you successfully ran tests with your changes locally?

### **After** these steps, you're ready to open a pull request.

    a. Give a descriptive title to your PR.
Support backend download feature
Example code:
```python
app = pc.App(state=State)
app.add_page(index)
app.add_download_file("/test", "ws/pynecone/download/test.JPG")
app.compile()
```
You can download file from localhost:8000/test and localhost:3000/test

    b. Describe your changes.

Use fastAPI to download file.

    c. Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

close #909

### CODE REVIEW

**Comment by TaiJuWu** [comment]
Is there method to check download file feature working correctly?

File: pynecone/app.py [hunk 1]
```diff
@@ -316,6 +321,34 @@ def get_load_events(self, route: str) -> List[EventHandler]:
...
+            raise ValueError(f"{download_url} already is used.")
+
+        self.backend_router.add(download_url)
+
+        @self.api.get(f"{download_url}")
+        async def response():
+            return FileResponse(
+                path=file_path,
+                filename=os.path.split(file_path)[1],
+                media_type="text/mp4",
```
Comment (masenf) [inline pynecone/app.py:347]: this would need to be parametrized and plumbed through

File: pynecone/app.py [hunk 2]
```diff
@@ -316,6 +321,34 @@ def get_load_events(self, route: str) -> List[EventHandler]:
...
+        if download_url in self.backend_router:
+            raise ValueError(f"{download_url} already is used.")
+
+        self.backend_router.add(download_url)
+
+        @self.api.get(f"{download_url}")
+        async def response():
+            return FileResponse(
+                path=file_path,
+                filename=os.path.split(file_path)[1],
```
Comment (masenf) [inline pynecone/app.py:346]: why `[1]`? Seems like you would want `[-1]` to grab the right-most element of the list. At any rate, I'd prefer to see `pathlib.Path(file_path).name` here instead.

File: pynecone/app.py [hunk 3]
```diff
@@ -316,6 +321,34 @@ def get_load_events(self, route: str) -> List[EventHandler]:
...
+            download_url: The route to get the file.
+            file_path: The file from host.
+
+        Returns:
+            Respone a file from server.
+
+        Raises:
+            ValueError: download_url already is used.
+        """
+        if download_url in self.backend_router:
```
Comment (masenf) [inline pynecone/app.py:337]: probably also need to check in `self.pages` for practical purposes.

**Review by masenf — CHANGES_REQUESTED** [review]
it needs to be really clear that the download files are on the backend server (port 8080 unless changed).

there should be a component that makes it easy to root-relative link to a downloadable file, by dereferencing the `api_url` config value. as it is now, there's no reliable way for the frontend to link to a downloadable file.

it might be nicer to leverage the `static_files` feature of `FastAPI`.

**Comment by TaiJuWu** [comment]
> it needs to be really clear that the download files are on the backend server (port 8080 unless changed).
> 
> there should be a component that makes it easy to root-relative link to a downloadable file, by dereferencing the `api_url` config value. as it is now, there's no reliable way for the frontend to link to a downloadable file.
> 
> it might be nicer to leverage the `static_files` feature of `FastAPI`.



> it needs to be really clear that the download files are on the backend server (port 8080 unless changed).
> 
> there should be a component that makes it easy to root-relative link to a downloadable file, by dereferencing the `api_url` config value. as it is now, there's no reliable way for the frontend to link to a downloadable file.
> 
> it might be nicer to leverage the `static_files` feature of `FastAPI`.

I have some questions about your suggestion.

1. Im not sure your 8080 is default backend server(8000) or not?

2. I use the method of api.get to download the file, so it will follow the setting of the config file to mount the donwload file url with api_url(default 8000).

3. I agree to there is not frontend to link download file but Im not sure it is necessary or not. What do you think?

4. Is fastAPI static file https://fastapi.tiangolo.com/tutorial/static-files/  ?

File: pynecone/app.py (hunk 3 above)
Comment (TaiJuWu) [reply pynecone/app.py:337]: pages seems to be used for frontend so I don't check this part.
If Im wrong, please tell me.

File: pynecone/app.py (hunk 2 above)
Comment (TaiJuWu) [reply pynecone/app.py:346]: I agree with you. Thanks for your suggestion.

**Comment by masenf** [comment]
1. Yea I typo'd, default is 8000
2. I think this is fine
3. The issue arises when the frontend code wants to link to a downloadable file, there isn't a easy way for frontend code to know what URL prefix is needed to access backend.
4. Yep that's the API I was referring to. My understanding is that serving files via static_files would be more efficient for the server.

File: pynecone/app.py (hunk 3 above)
Comment (masenf) [reply pynecone/app.py:337]: i think a common pattern for deployments is to use a reverse proxy that forwards requests to either frontend or backend based on the path, so there's a possibility of conflict in these cases. maybe it's not a big deal though.

**Comment by TaiJuWu** [comment]
> * Yea I typo'd, default is 8000
> * I think this is fine
> * The issue arises when the frontend code wants to link to a downloadable file, there isn't a easy way for frontend code to know what URL prefix is needed to access backend.
> * Yep that's the API I was referring to. My understanding is that serving files via static_files would be more efficient for the server.

I have tried to automatically create backend download file by frontend component but it's very difficult so I give up and only implement the backend download file.

Static files seem not working because it will only show image/video on browsers instead of download file.

File: pynecone/app.py (hunk 3 above)
Comment (TaiJuWu) [reply pynecone/app.py:343]: I used self.add _page to redirect frontend to backend and there will check self.page.

**Comment by Justin900429** [comment]
Hi, I have a few concerns regarding the new download support.

Firstly, it seems that the default media type is set to `mp4`. However, in your example, you didn't specify the media type and instead returned an image. I'm wondering if this could have any impact on the content file in the response header.

```python
# Provide the media type correctly
app.add_download_file("/test", "ws/pynecone/download/test.JPG", media_type="image/jpeg")
```

Secondly, I suggest considering removing the default value for the `media_type` parameter. By doing so, users will be required to explicitly specify the media type, which can help prevent any unintended side effects.

These adjustments should enhance clarity and ensure that users are aware of the media type being used and can make a conscious decision based on their specific requirements.

**Pull request closed without merging**